dotNET: added page for nested table support (#6226)
EnterpriseDB/docs: product_docs/docs/net_connector/8.0.5.1/using_nested_table_types.mdx

---
title: "Using nested tables"

---

> EDB Postgres Advanced Server supports nested table collection types created with `CREATE TYPE ... AS TABLE OF` statements. The EDB .NET Connector supports output parameters declared as nested tables out of the box, whether they are free-standing types or declared inside packages.

Reviewer: ```suggestion
EDB Postgres Advanced Server supports nested table collection types created with `CREATE TYPE ... AS TABLE OF` statements. The EDB .NET Connector supports output parameters declared as nested tables out of the box, whether free-standing types or declared inside packages.
```

## Nested table types mapping

> Nested table types are mapped to `List<object>`s in C#, as it is preferred over `ArrayList`. These lists contain as many elements as the nested table type’s rows. The nested table items are translated to be compatible with the C# application using the following rules:

Reviewer: ```suggestion
Nested table types are mapped to `List<object>`s in C#, as it is preferred over `ArrayList`. These lists contain as many elements as the nested table type's rows. The nested table items are translated to be compatible with the C# application using the following rules:
```

> - The connector resolves all elements of a nested table into `List<object>` elements of a single object type, making the element types consistent in C#.

Author: Hmmm let me use an example to see if I got it: 
So let's say in a nested table called "order", we have object types like `first_name`, `last_name`, and `orderID`. 
When .NET makes a call, it transforms/maps those object types to `List<object>`s (a list of objects). BUT the list items will keep the `first_name`, `last_name`, and `orderID` "metadata"?

> - The connector resolves all elements of a nested table into `List<object>` elements of a single object type, making the element types consistent in C#.

Author: Gotcha, thanks so much for the explanation! 🙏  In that case, what do you think of this: 
```suggestion
- The connector resolves all items of a nested table into a `List<object>` in C# while maintaining and converting each item's underlying type. For example, `[text1, text1, num1]` becomes `[string1, string2, decimal1]`.
```

> - The connector resolves all elements of a nested table into `List<object>` elements of a single object type, making the element types consistent in C#.

Reviewer 2: ```suggestion
- The connector resolves all rows of a nested table into a `List<object>` in C# while maintaining and converting each column underlying type. For example, a `[text1, text1, num1]` row will be resolved as a `[string, string, decimal]` item in the list.
```

I've replaced "item" by row:

- table of type has rows
- each row has columns
- List<object> has items

If you agree, we can replace that in next bullet list

- If the nested type IS TABLE OF a domain type (int, varchar, decimal, etc.), all the rows will be their C# counterpart according to the [Supported Types and their Mappings](https://www.npgsql.org/doc/types/basic.html#read-mappings).

- If the nested type IS TABLE OF a record or composite type not mapped to a C# class, all rows become a nested List<object> containing as many elements as the record or composite fields, with proper type translation.

- If the nested type IS TABLE OF a record or composite type mapped to a C# class (for example, MyComposite), all rows will be MyComposite instances.

> - The connector resolves all elements of a nested table into `List<object>` elements of a single object type, making the element types consistent in C#.

Author: Awesome, I made the corrections and I think we are good to go 🥳

> - The connector resolves all rows of a nested table into a `List<object>` in C# while maintaining and converting each column's underlying type. For example, a `[text1, text2, num1]` row will be resolved as a `[string, string, decimal]` item in the list.

Reviewer: ```suggestion
- The connector resolves all nested table rows into a `List<object>` in C# while maintaining and converting each column's underlying type. For example, a `[text1, text2, num1]` row will be resolved as a `[string, string, decimal]` item in the list.
```

> - If the nested type `IS TABLE OF` a domain type (int, varchar, decimal, etc.), all the rows will be their C# counterpart according to the [Supported Types and their Mappings](https://www.npgsql.org/doc/types/basic.html#read-mappings).

Reviewer: ```suggestion
- If the nested type `IS TABLE OF` is a domain type (int, varchar, decimal, etc.), all the rows will be their C# counterpart according to the [Supported Types and their Mappings](https://www.npgsql.org/doc/types/basic.html#read-mappings).
```

- If the nested type `IS TABLE OF` a record or composite type **not mapped** to a C# class, all rows become a nested List containing as many elements as the record or composite fields, with proper type translation.

> - If the nested type `IS TABLE OF` a record or composite type **mapped** to a C# class (for example, `MyComposite`), all rows will be `MyComposite` instances.

Reviewer: ```suggestion
- If the nested type `IS TABLE OF` is a record or composite type **mapped** to a C# class (for example, `MyComposite`), all rows will be `MyComposite` instances.
```

## Example: Retrieving nested table output parameter

This program:

- Creates a package with a nested `emp_tbl_typ` table type of `emp_rec_typ`. This package has a stored procedure that fills the nested table output parameter.

- Maps the nested table type to a C# class via `MapComposite<>`.

- Executes and displays the results.

!!!note
   Always provide type names in lowercase.
!!!

### Program example

Create an empty console program and paste the following code.

> ```text

Reviewer 2: can you make this ```csharp
for better syntax coloring ?

"                    BEGIN \n" +
"                       emp_tbl := emp_tbl_typ();  \n" +
"                       FOR r_emp IN emp_cur LOOP  \n" +
"                          i := i + 1;  \n" +
"                          emp_tbl.EXTEND;  \n" +
"                          emp_tbl(i) := r_emp;  \n" +
"                       END LOOP;  \n" +
"                    END nestedTableExtendTest; \n" +
"                    END pkgExtendTest; \n";
        using (var com = new EDBCommand(createPackageBody, connection) { CommandType = CommandType.Text })
        {
            com.ExecuteNonQuery();
        }

        connection.ReloadTypes();
    }

    static void Cleanup(EDBConnection connection)
    {
        var dropPackageBody = "DROP PACKAGE BODY pkgExtendTest";
        var dropPackage = "DROP PACKAGE pkgExtendTest";

        using (var com = new EDBCommand(dropPackageBody, connection) { CommandType = CommandType.Text })
        {
            com.ExecuteNonQuery();
        }
        using (var com = new EDBCommand(dropPackage, connection) { CommandType = CommandType.Text })
        {
            com.ExecuteNonQuery();
        }
    }
}
```

The output should look like this: 

> ```text

Reviewer 2: Here text is OK
